contracts/finance/ProtocolStaking.sol

```diff
@@ -0,0 +1,259 @@
+    using Checkpoints for Checkpoints.Trace208;
+    using SafeERC20 for IERC20;
+
+    struct UserStakingInfo {
+        uint256 rewardsPerUnitPaid;
+        uint256 rewards;
+    }
+
+    bytes32 private constant OPERATOR_ROLE = keccak256(bytes("operator-role"));
+    address private _stakingToken;
+    uint256 private _totalStakedWeight;
+    uint256 private _lastUpdateTimestamp;
+    uint256 private _rewardsPerUnit;
+    uint256 private _rewardRate;
+    uint256 private _unstakeCooldownPeriod;
+    mapping(address => UserStakingInfo) private _userStakingInfo;
+    mapping(address => Checkpoints.Trace208) private _unstakeRequests;
+    mapping(address => uint256) private _totalReleased;
+    mapping(address => address) private _rewardsRecipient;
+
+    event OperatorAdded(address operator);
+    event OperatorRemoved(address operator);
+    event TokensStaked(address operator, uint256 amount);
+    event TokensUnstaked(address operator, uint256 amount);
+    event RewardRateSet(uint256 rewardRate);
+    event UnstakeCooldownPeriodSet(uint256 unstakeCooldownPeriod);
+    event RewardsRecipientSet(address indexed account, address indexed recipient);
+
+    error InvalidAmount();
+    error OperatorAlreadyExists(address operator);
+    error OperatorDoesNotExist(address operator);
+    error TransferDisabled();
+
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        string memory name,
+        string memory symbol,
+        string memory version,
+        address stakingToken_,
+        address governor
+    ) public virtual initializer {
+        __AccessControlDefaultAdminRules_init(0, governor);
+        __ERC20_init(name, symbol);
+        __EIP712_init(name, version);
+        _stakingToken = stakingToken_;
+        _rewardsPerUnit = 1; // initialize rewards per unit
+    }
+
+    function stake(uint256 amount) public virtual {
+        _stake(amount);
+    }
+
+    function unstake(uint256 amount) public virtual {
+        _unstake(amount);
+    }
+
+    function release() public virtual {
+        uint256 totalAmountCooledDown = _unstakeRequests[msg.sender].upperLookup(Time.timestamp());
+        uint256 amountToRelease = totalAmountCooledDown - _totalReleased[msg.sender];
+        _totalReleased[msg.sender] = totalAmountCooledDown;
+        if (amountToRelease > 0) {
+            IERC20(stakingToken()).safeTransfer(msg.sender, amountToRelease);
+        }
+    }
+
+    function earned(address account) public view virtual returns (uint256) {
+        UserStakingInfo memory userInfo = _userStakingInfo[account];
+        if (userInfo.rewardsPerUnitPaid == 0) {
+            return userInfo.rewards;
+        }
+        return (weight(balanceOf(account)) * (_rewardsPerUnit - userInfo.rewardsPerUnitPaid)) / 1e18 + userInfo.rewards;
+    }
+
+    /// @dev Claim staking rewards for `account`.
+    function claimRewards(address account) public virtual {
```

Comment:
It looks anyone can force the account to receive rewards on his `account` if the account has never set reward recipient with `setRewardsRecipient(..)` (eventually possible but always possible by front running `setRewardsRecipient(..)` with a claim). The receiver should have full control on that (eg: `recipient` initialized when staking) or instead should get a deterministic behavior (such as all rewards goes to account).

Reply:
The staker can opt to set their reward address before staking such that all rewards go to that address. Otherwise the default behavior of sending the rewards to the account its self is used. I don’t see the issue.